openrarity/collection/collection.py

```diff
@@ -0,0 +1,337 @@
+
+    @property
+    def tokens(self) -> dict[TokenId, RawToken]:
+        return self._tokens
+
+    @property
+    def total_supply(self) -> int:
+        # SemiFungible needs to sum the supply of each token
+        if isinstance(self._token_supply, dict):
+            return sum(self._token_supply.values())
+
+        return self._token_supply
+
+    @property
+    def attribute_statistics(self) -> AttributesStatistics:
+        if not self._attribute_statistics:
+            raise AttributeError(
+                f"Please run '{repr(self)}.rank_collection()' to view this property"
+            )
+        return self._attribute_statistics
+
+    @property
+    def entropy(self) -> float | None:
+        return self._entropy
+
+    @property
+    def token_statistics(self) -> list[TokenStatistic]:
+        if not self._token_statistics:
+            raise AttributeError(
+                f"Please run '{repr(self)}.rank_collection()' to view this property"
+            )
+        return self._token_statistics
+
+    @property
+    def ranks(self) -> list[RankedToken]:
+        if not self._ranks:
+            raise AttributeError(
+                f"Please run '{repr(self)}.rank_collection()' to view this property"
+            )
+        return self._ranks
+
+    @property
+    def checksum(self) -> str:
+        if not self._ranks_checksum:
+            raise AttributeError(
+                f"Please run '{repr(self)}.rank_collection()' to view this property"
+            )
+        return self._input_checksum + self._ranks_checksum
+
+    @overload
+    def rank_collection(  # type: ignore
+        self,
+        rank_by: tuple[
+            Literal[
+                "metric.probability",
+                "metric.information",
+                "metric.information_entropy",
+                "metric.unique_trait_count",
+                "metric.max_trait_information",
+            ],
+            ...,
+        ] = ("metric.unique_trait_count", "metric.information_entropy"),
+        return_ranks: Literal[True] = True,
+    ) -> list[RankedToken]:
+        ...
+
+    @overload
+    def rank_collection(  # type: ignore
+        self,
+        rank_by: tuple[
+            Literal[
+                "metric.probability",
+                "metric.information",
+                "metric.information_entropy",
+                "metric.unique_trait_count",
+                "metric.max_trait_information",
+            ],
+            ...,
+        ] = ("metric.unique_trait_count", "metric.information_entropy"),
+        return_ranks: Literal[False] = False,
+    ) -> None:
+        ...
+
+    def rank_collection(
+        self,
+        rank_by: tuple[
+            Literal[
+                "metric.probability",
+                "metric.information",
+                "metric.information_entropy",
+                "metric.unique_trait_count",
+                "metric.max_trait_information",
+            ],
+            ...,
```

Comment:
Not familiar with ... syntax, what is it doing?

Reply:
The ... in tuple[] just means that it can be of arbitrary length. So specifically the above typing says that it is a tuple with any number of the above literal values.